src/riscv/lib/src/jit/builder/sequence.rs

```diff
@@ -278,173 +384,72 @@ impl<'jit, MC: MemoryConfig> SequenceBuilder<'jit, MC> {
             self.builder.ins().return_(&[steps]);
         }
 
-        let mut peekable_instrs = instrs.iter().enumerate().peekable();
-
-        if let Some((_, first_instr)) = peekable_instrs.peek() {
-            // Hook up the entry block to the first instruction.
-            self.builder.switch_to_block(self.entry_block);
-            first_instr.build_run(&mut self.builder);
-        }
-
-        while let Some((instr_index, instr)) = peekable_instrs.next() {
-            // This step populates the hook block of each outcome of the instruction as they have not been
-            // handled yet. The hook blocks are created during instruction building, but the control flow
-            // transitions are only handled here, once all instructions have been built.
-            match instr.outcomes() {
-                InstructionOutcomes::Next { hook } => {
-                    let next_instr = peekable_instrs.peek().map(|(_, instr)| *instr);
-                    self.handle_next_outcome(*hook, instr_index, instr, next_instr, exit_block);
-                }
-
-                InstructionOutcomes::Relative { offset, hook } => {
-                    self.handle_relative_outcome(offset, *hook, instr_index, instr, exit_block);
-                }
-
-                InstructionOutcomes::Absolute { destination, hook } => {
-                    self.handle_absolute_outcome(*destination, *hook, instr_index, exit_block);
-                }
+        let node_infos: Vec<NodeInfo<OutcomeData, Block>> = instrs
+            .iter()
+            .enumerate()
+            .map(|(idx, instr)| {
+                NodeInfo::<OutcomeData, Block>::from_lowered_instruction(instr, idx == 0)
+            })
+            .collect();
+
+        let cfg = ControlFlowGraph::<OutcomeData, Block>::new(node_infos.iter());
+
+        let (step_updates, exit_deltas) = cfg.find_step_counter_updates();
+        let bc_locations = cfg.find_budget_check_locations();
+        let sequence_budget = cfg.annotate_budget_checks(bc_locations, &exit_deltas);
```

Comment:
I think `annotate_budget_checks` always takes the `bc_locations` obtained just beforehand. Would it make sense for `annotate_budget_checks` to do this internally instead?